Allow a percentage of traffic to be proxied to S3 via Nginx

floehopper opened the pull request:
The motivation behind this commit is to allow us to incrementally roll-out the use of the proxying to S3 via Nginx mechanism. Although we've done a bunch of performance testing and analysis of the production logs, it's difficult to be certain that the production stack will be able to handle all its usual traffic when this mechanism is enabled. By only enabling it gradually we hope to observe any problems before they become catastrophic.

The `PROXY_ALL_ASSET_REQUESTS_TO_S3_VIA_NGINX` environment variable has been renamed to `PROXY_PERCENTAGE_OF_ASSET_REQUESTS_TO_S3_VIA_NGINX` and this should be set to an integer value between 0 and 100. We will need to set appropriate values for each environment via `govuk-puppet`.

This addresses the Asset Manager aspect of #151, but we will need to set values for this new environment variable via `govuk-puppet`.

I ended up doing quite a bit of tidying of the code prior to making the actual change; the latter is the final commit. Other changes include:

* Making other environment variable names consistent (I don't think we've used them anywhere yet, so this should be safe)
* Making code & specs more consistent.
* Improving test coverage and reducing combinatorial explosion in specs.
* Improve documentation about feature flags.

I think there's a case to be made to extract the response strategy selection logic and the response strategies themselves into their own classes - this would make the `MediaController#download` action easier to understand and easier to test. I did make a start on doing this, but decided it was a step too far for this PR. It may also be moot if we remove some of the response strategies in the near future.

floehopper commented:
@chrisroos: If you're going to do the `govuk-puppet` work related to this, it might make sense for you to review it...

chrisroos commented:
The body of the commit titled "Improve consistency & naming in MediaController spec" contains "* Avoid use of "should" in example names". Is there a styleguide or similar that explains we shouldn't be using `should` in test names?

chrisroos commented:
The first sentence of the commit titled, "Make order of asset response strategies more logical", reads "I've changed the order of the conditions in the `if/elsif/else` statement to a more logical.". Is there a word missing?

floehopper commented:
@chrisroos:

Thanks for reviewing.

> The body of the commit titled "Improve consistency & naming in MediaController spec" contains "* Avoid use of "should" in example names". Is there a styleguide or similar that explains we shouldn't be using should in test names?

There is in the [`RSpec/ExampleWording` cop][1] in the `rubocop-rspec` gem which I think we should be using (see #92). The cop documentation refers to [this page][2].

> The first sentence of the commit titled, "Make order of asset response strategies more logical", reads "I've changed the order of the conditions in the `if/elsif/else` statement to a more logical.". Is there a word missing?

Yes. I'll fix it. It should say something more like: "I've changed the conditions in the `if/elsif/else` statement to be in a more logical order."

[1]: https://github.com/backus/rubocop-rspec/blob/79818d3c184552e039f6b66da10c2e641eb00cfd/lib/rubocop/cop/rspec/example_wording.rb#L6-L22
[2]: http://www.betterspecs.org/#should

floehopper commented:
Rebased against `master`, fixed commit wording, and force-pushed in preparation for merging.
